### PR TITLE
Volunteer Schedule Page

### PR DESCRIPTION
**Pull request by jjessieshang**
Screenshot 2024-11-23 at 1 41 16 PM
Screenshot 2024-11-23 at 1 40 54 PM

# **PR Description: Completed Shift Schedule Page**

This PR implements the **Shift Schedule** page, allowing volunteers to:  
1. View their assigned shifts.  
2. See open coverage requests from other volunteers.  
3. Track their own coverage requests.  

---

## **Database Changes**  
1. **`checked_in` Field Added**:  
   - Added to the `shifts` table to track volunteer check-ins.  

2. **`pending_shift_coverage` Table Added**:  
   - Tracks pending coverage requests and their fulfillments.

---

## **New Endpoints**  
1. **`/request-to-cover-shift`**  
   - Allows volunteers to request coverage for open shifts.  

2. **`/volunteer-month`**  
   - Retrieves all shifts viewable to a volunteer for a given month/year.  
   - Includes:  
     - **`my-shifts`**: Assigned shifts.  
     - **`coverage`**: Open shifts available for coverage.  
     - **`my-coverage-requests`**: Coverage requests made by the logged-in volunteer.  
   - Returns shift details such as date, time, class, duration, and coverage status.  

---

## **Frontend Updates**  
1. Implemented the **Shift Schedule Page** UI/UX.  
- VolunteerSchedule page
- DateToolbar component
- ShiftCard component
- ShiftStatusToolbar

3. **Volunteer Authentication**:  
   - Stores logged-in, approved volunteer IDs in `localStorage` for API routing and access control.  

TODO
- shift details side panel
---

### CODE REVIEW

File: backend/src/models/shiftModel.ts [hunk 1]
```diff
@@ -76,4 +76,45 @@ export default class ShiftModel {
...
+              const query = `
+                  CALL GetShiftsByVolunteerIdAndMonth(?, ?, ?);
+              `;
+              
+              const values = [volunteer_id, month, year];
+              connection.query(query, values, (error: any, results: any) => {
+                  if (error) {
+                      return reject({
+                          status: 500,
+                          message: `An error occurred while executing the query: ${error}`
```
Comment (alecishuman) [inline backend/src/models/shiftModel.ts:92]: Just gonna add this here for now, but I'm pretty sure this error handling doesn't make sense. Idk why its all over our project, but putting `error` here makes more sense. Not a problem particularly, but maybe smt to note for future endpoints.

**Review by alecishuman — APPROVED** [review]
Looks good to me :)

File: backend/src/models/shiftModel.ts (hunk 1 above)
Comment (alecishuman) [reply backend/src/models/shiftModel.ts:92]: Don't need to change if its too annoying

Comment (jjessieshang) [reply backend/src/models/shiftModel.ts:92]: noted :)

**Pull request merged**